scalding-core/src/main/scala/com/twitter/scalding/Execution.scala

```diff
@@ -490,20 +521,55 @@ object Execution {
    * are based on on this one. By keeping the Pipe and the Sink, can inspect the Execution
    * DAG and optimize it later (a goal, but not done yet).
    */
   private case class WriteExecution[T](head: ToWrite, tail: List[ToWrite], fn: (Config, Mode) => T) extends Execution[T] {
-    def runStats(conf: Config, mode: Mode, cache: EvalCache)(implicit cec: ConcurrentExecutionContext) =
-      cache.getOrElseInsert(this,
-        for {
-          flowDef <- toFuture(Try { val fd = new FlowDef; (head :: tail).foreach(_.write(conf, fd, mode)); fd })
-          _ = FlowStateMap.validateSources(flowDef, mode)
-          jobStats <- cache.runFlowDef(conf, mode, flowDef)
-          _ = FlowStateMap.clear(flowDef)
-        } yield (fn(conf, mode), ExecutionCounters.fromJobStats(jobStats)))
+
+    /* Run a list of ToWrite elements */
+    private[this] def scheduleToWrites(conf: Config,
+      mode: Mode,
+      cache: EvalCache,
+      head: ToWrite,
+      tail: List[ToWrite])(implicit cec: ConcurrentExecutionContext): Future[ExecutionCounters] = {
+      for {
+        flowDef <- toFuture(Try { val fd = new FlowDef; (head :: tail).foreach(_.write(conf, fd, mode)); fd })
+        _ = FlowStateMap.validateSources(flowDef, mode)
+        jobStats <- cache.runFlowDef(conf, mode, flowDef)
+        _ = FlowStateMap.clear(flowDef)
+      } yield (ExecutionCounters.fromJobStats(jobStats))
+    }
+
+    // We look up to see if any of our ToWrite elements have already been ran
+    // if so we remove them from the cache.
+    // Anything not already ran we run as part of a single flow def, using their combined counters for the others
+    def runStats(conf: Config, mode: Mode, cache: EvalCache)(implicit cec: ConcurrentExecutionContext) = {
+      val cacheLookup: List[(ToWrite, Either[Promise[ExecutionCounters], Future[ExecutionCounters]])] = (head :: tail).map{ tw => (tw, cache.getOrLock(tw)) }
+      val (weDoOperation, someoneElseDoesOperation) = cacheLookup.partition(_._2.isLeft)
+
+      val localFlowDefCountersFuture: Future[ExecutionCounters] =
+        weDoOperation match {
+          case all @ (h :: tail) =>
+            val futCounters: Future[ExecutionCounters] = scheduleToWrites(conf, mode, cache, h._1, tail.map(_._1))
+            // Complete all of the promises we put into the cache
+            // with this future counters set
+            weDoOperation.foreach {
+              case (toWrite, eitherP) =>
+                eitherP.left.get.completeWith(futCounters)
+            }
+            futCounters
+          case Nil => Future.successful(ExecutionCounters.empty) // No work to do, provide a fulled set of 0 counters to operate on
+        }
+
+      failFastZip(failFastSequence(someoneElseDoesOperation.map(_._2.right.get)), localFlowDefCountersFuture).map {
+        case (lCounters, fdCounters) =>
+          val summedCounters: ExecutionCounters = Monoid.sum(fdCounters :: lCounters)
+          (fn(conf, mode), summedCounters)
+      }
+    }
 
```

Comment:
Does it avoid anything to inline Map instead of having the MappedExecution? if its in the presentation function we won't cache it anyway. (Indeed if the map function is expensive we would want to cache it as we do MappedExecutions right?)